File: rust/src/solutions/remove_linked_list_elements.rs
// LeetCode 203. Remove Linked List Elements in Rust

use super::Solution;
use crate::common::ListNode;

// impl Solution {
//     pub fn remove_elements(head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
//         let mut dummy = Box::new(ListNode::new(0));
//         dummy.next = head;
//         let mut ptr = &mut dummy;
//         while ptr.next.is_some() {
//             if ptr.next.as_ref().unwrap().val == val{
//                 ptr.next = ptr.next.as_mut().unwrap().next.take();
//             } else {
//                 ptr = ptr.next.as_mut().unwrap();
//             }
//         }
//         dummy.next
//     }
// }

impl Solution {
    pub fn remove_elements(head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        let mut head = head;
        let mut ans = None;
        let mut tail = &mut ans;
        while let Some(mut node) = head.take() {
            head = node.next.take();
            if node.val != val {
                tail = &mut tail.insert(node).next;
            }
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;
    use crate::common::{list_create, list_equal};

    #[test]
    fn example1() {
        let head = list_create(&[1, 2, 6, 3, 4, 5, 6]);
        let val = 6;
        let ans = list_create(&[1, 2, 3, 4, 5]);
        assert!(list_equal(&Solution::remove_elements(head, val), &ans));
    }
    #[test]
    fn example2() {
        let head = list_create(&[]);
        let val = 1;
        let ans = list_create(&[]);
        assert!(list_equal(&Solution::remove_elements(head, val), &ans));
    }
    #[test]
    fn example3() {
        let head = list_create(&[7, 7, 7, 7]);
        let val = 7;
        let ans = list_create(&[]);
        assert!(list_equal(&Solution::remove_elements(head, val), &ans));
    }
}

File: cpp/src/solutions/remove_linked_list_elements.cpp
// LeetCode 203. Remove Linked List Elements

#include "cpp/src/common/common.h"

class Solution {
 public:
  ListNode* removeElements(ListNode* head, int val) {
    ListNode dummy(0, head);
    ListNode* ptr = &dummy;
    while (ptr->next != nullptr) {
      if (ptr->next->val == val) {
        auto tmp = ptr->next;
        ptr->next = tmp->next;
        delete tmp;
      } else {
        ptr = ptr->next;
      }
    }
    return dummy.next;
  }
};

TEST(removeElements, example1) {
  ListNode* head = list_create({1, 2, 6, 3, 4, 5, 6});
  int val = 6;
  ListNode* ans = list_create({1, 2, 3, 4, 5});
  EXPECT_TRUE(list_equal(Solution().removeElements(head, val), ans));
}

TEST(removeElements, example2) {
  ListNode* head = list_create({});
  int val = 1;
  ListNode* ans = list_create({});
  EXPECT_TRUE(list_equal(Solution().removeElements(head, val), ans));
}

TEST(removeElements, example3) {
  ListNode* head = list_create({7, 7, 7, 7});
  int val = 7;
  ListNode* ans = list_create({});
  EXPECT_TRUE(list_equal(Solution().removeElements(head, val), ans));
}
